### PR TITLE
Protection against SQL injections

### PR DESCRIPTION
**Pull request by PatrOlen**
Implemented methods of prevention against SQL injections:

- Use of placeholders for parametrized queries
image
`?` means that we are using a placeholder for a parameter in a query. This parameter must be supplied in the querySQL() function as a second argument. When parametrizing column, table, or variable names, `??` should be used instead of `?`.
This way malicious code will be escaped.

- Added RegEx and length validation for user input.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,6 +18,7 @@
     "dotenv": "^16.0.3",
     "i18next": "^22.4.9",
     "mysql": "^2.18.1",
```
Comment (krzotki) [inline package.json:20]: This could be removed if unused

File: pages/api/auth/[...nextauth].ts [hunk 2]
```diff
@@ -63,13 +70,12 @@ export const authOptions = {
         }>;
 
         const match = await bcrypt.compare(
-          credentials?.password,
+          credentials?.password!,
```
Comment (krzotki) [inline pages/api/auth/[...nextauth].ts:73]: Password should be defined at this point

File: pages/list.tsx [hunk 3]
```diff
@@ -300,18 +300,18 @@ export async function getServerSideProps({
   if (session) {
     // @ts-ignore
     const userid = session?.user?.id;
-    favorites = (await getFavoriteCryptos(userid)).map(
+    favorites = (await getFavoriteCryptos(userid!)).map(
```
Comment (krzotki) [inline pages/list.tsx:303]: UserId will be undefined if user is not logged in.

File: pages/list.tsx [hunk 4]
```diff
@@ -300,18 +300,18 @@ export async function getServerSideProps({
...
       (row: any) => row.Cf_CryptoId
     );
 
     thresholds = Object.fromEntries(
-      (await getCryptothresholds(userid)).map((row: any) => [
+      (await getCryptothresholds(userid!)).map((row: any) => [
```
Comment (krzotki) [inline pages/list.tsx:308]: Same as above

File: pages/list.tsx [hunk 5]
```diff
@@ -300,18 +300,18 @@ export async function getServerSideProps({
...
         row.Cn_CryptoId,
         row.Cn_Treshold,
       ])
     );
 
-    favoriteMarket = await getFavoriteMarket(userid);
+    favoriteMarket = await getFavoriteMarket(userid!);
```
Comment (krzotki) [inline pages/list.tsx:314]: ^^

**Review by krzotki — COMMENTED** [review]
Make sure every feature works for unlogged and logged works. Do this in production mode - npm start

**Comment by PatrOlen** [comment]
image
Please change NEXTAUTH_URL variable in your .env file, from `localhost` to `127.0.0.1`. While it may seem like it's the same thing, the `localhost causes fetching errors in production.

**Review by krzotki — APPROVED** [review]
🥃

**Pull request merged**